### PR TITLE
[skip] Fixes to keep pipelines consistency

### PR DESCRIPTION
**Pull request by kyfujisa**
As a fix to keep pipelines consistent, the first step.

## confirmation

test.sh
```bash
#mkdir -p $HOME/.ssh/ && echo -e "Host github.com\n\tStrictHostKeyChecking no\n" > ~/.ssh/config
#LAST_COMMIT=`git log -1 --pretty=%B`
LAST_COMMIT=$1
#if [ ! -z $CIRCLE_PULL_REQUEST ]; then
#  PR_NUM="${CIRCLE_PULL_REQUEST##*/}"
#  LAST_COMMIT=`curl -s ${GITHUB_API}repos/${REPO_NAME}/${IMAGE_NAME}/pulls/${PR_NUM} | jq .title`
#fi
touch ./.tag
# Diff to prevent pipeline re-runs.
#if VERSION=`git describe --abbrev=0 --tags` && [ ! -z "`git diff $VERSION`" -o -z "$VERSION" ]; then
#VERSION=`git describe --abbrev=0 --tags`
VERSION=$2
#if [ ! -z "`git diff $VERSION`" -o -z "$VERSION" ]; then
if [ -n "$3" ]; then
  VERSION=${VERSION:-'0.0.0'}
  MAJOR="${VERSION%%.*}"; MAJOR="${MAJOR#v}"; VERSION="${VERSION#*.}"
  MINOR="${VERSION%%.*}"; VERSION="${VERSION#*.}"
  PATCH="${VERSION%%.*}"; VERSION="${VERSION#*.}"
  case "$LAST_COMMIT" in
  '[patch]'* | '[PATCH]'*)
    PATCH=$((PATCH+1))
    echo "v$MAJOR.$MINOR.$PATCH" > ./.tag
    ;;
  '[minor]'* | '[MINOR]'*)
    MINOR=$((MINOR+1))
    echo "v$MAJOR.$MINOR.0" > ./.tag
    ;;
  '[major]'* | '[MAJOR]'*)
    MAJOR=$((MAJOR+1))
    echo "v$MAJOR.0.0" > ./.tag
    ;;
  # Don't release (nightly images only)
  '[skip]'* | '[SKIP]'*)
    echo "Skip release..."
    ;;
  # Don't want to run CI.
  '[skip ci]'* | '[ci skip]'*)
    echo "Will skip CI..."
    ;;
  *)
    echo "Required prefixes: [major]/[minor]/[patch]/[skip]/[skip ci]"
    exit 1
    ;;
  esac
else
  if [ -n "$VERSION" ]; then
    echo "No diff, skip tagging..."
  else
    echo "No diff, No tag. Tag it as \"v0.0.0\"."
    echo "v0.0.0" > ./.tag
  fi
fi

# print tag
cat ./.tag
rm ./.tag

exit 0
```
check
```bash
$ # tag check
$ sh test.sh '[patch] commit' v1.1.1 diff
v1.1.2
$ sh test.sh '[minor] commit' v1.1.1 diff
v1.2.0
$ sh test.sh '[major] commit' v1.1.1 diff
v2.0.0
$ sh test.sh '[skip] commit' v1.1.1 diff
Skip release...
$ sh test.sh '[skip ci] commit' v1.1.1 diff
Will skip CI...
$ sh test.sh 'no tag commit' v1.1.1 diff
Required prefixes: [major]/[minor]/[patch]/[skip]/[skip ci]
$ echo $?
1
$ # no version
$ sh test.sh '[patch] commit' '' 'diff'
v0.0.1
$ sh test.sh '[minor] commit' '' 'diff'
v0.1.0
$ sh test.sh '[major] commit' '' 'diff'
v1.0.0
$ # no diff
$ sh test.sh '[major] commit' 'v2.0.0' ''
No diff, skip tagging...
$ sh test.sh '[major] commit' '' ''
No diff, No tag. Tag it as "v0.0.0".
v0.0.0
$ sh test.sh 'no tag commit' 'v2.0.0' ''
No diff, skip tagging...
$ sh test.sh 'no tag commit' '' ''
No diff, No tag. Tag it as "v0.0.0".
v0.0.0
```

### CODE REVIEW

**Comment by kyfujisa** [comment]
I think this issue has something to do with the failure of the pipeline.
https://github.com/docker/cli/issues/2533

**Comment by WindzCUHK** [comment]
> I think this issue has something to do with the failure of the pipeline.

@kyfujisa please update the docker version in pipeline.
referenece: https://github.com/yahoojapan/garm/pull/39/commits/79472244b64c251ac24111c104d2330b785c372f

**Comment by WindzCUHK** [comment]
try to simply the logic as below

`case.sh`
```bash
#!/bin/sh

# LAST_COMMIT="$(git log -1 --pretty=%B)"
LAST_COMMIT="$1"
# if [ ! -z $CIRCLE_PULL_REQUEST ]; then
#   PR_NUM="${CIRCLE_PULL_REQUEST##*/}"
#   LAST_COMMIT=`curl -s ${GITHUB_API}repos/${REPO_NAME}/${IMAGE_NAME}/pulls/${PR_NUM} | jq .title`
# fi
touch ./.tag

# VERSION="$(git describe --abbrev=0 --tags)"
VERSION="$2"
# if [ -n "$(git diff $VERSION)" ]; then
if [ -n "$3" ]; then
  VERSION=${VERSION:-'v0.0.0'}
  MAJOR="${VERSION%%.*}"; MAJOR="${MAJOR#v}"; VERSION="${VERSION#*.}"
  MINOR="${VERSION%%.*}"; VERSION="${VERSION#*.}"
  PATCH="${VERSION%%.*}"; VERSION="${VERSION#*.}"

  case "$LAST_COMMIT" in
    '[patch]'*)
      PATCH=$((PATCH+1))
      echo "v$MAJOR.$MINOR.$PATCH" > ./.tag
      ;;
    '[minor]'*)
      MINOR=$((MINOR+1))
      echo "v$MAJOR.$MINOR.0" > ./.tag
      ;;
    '[major]'*)
      MAJOR=$((MAJOR+1))
      echo "v$MAJOR.0.0" > ./.tag
      ;;
    '[skip]'*)
      echo "Skip release..."
      ;;
    '[skip ci]'* | '[ci skip]'*)
      echo "Will skip CI..."
      ;;
    *)
      echo "Required prefixes: [major]/[minor]/[patch]/[skip]/[skip ci]"
      exit 1
  esac
else
  echo "NO diff, skip tagging..."
fi

cat ./.tag
rm ./.tag

exit 0
```

output
```bash
$ sh case.sh '[patch] commit' v1.1.1 diff
v1.1.2
$ sh case.sh '[minor] commit' v1.1.1 diff
v1.2.0
$ sh case.sh '[major] commit' v1.1.1 diff
v2.0.0
$ sh case.sh '[skip] commit' v1.1.1 diff
Skip release...
$ sh case.sh '[skip ci] commit' v1.1.1 diff
Will skip CI...
$ sh case.sh '[-] commit' v1.1.1 diff
Required prefixes: [major]/[minor]/[patch]/[skip]/[skip ci]
$ echo $?
1
$ sh case.sh '[patch] commit' v1.1.1 ""
NO diff, skip tagging...
$ sh case.sh '[patch] commit' '' 'diff'
v0.0.1
$ sh case.sh '[minor] commit' '' 'diff'
v0.1.0
$ sh case.sh '[major] commit' '' 'diff'
v1.0.0
$ sh case.sh '[major] commit' 'v2.0.2' 'diff'
v3.0.0
$ sh case.sh '[minor] commit' 'v2.0.2' 'diff'
v2.1.0
$ sh case.sh '[patch] commit' 'v2.0.2' 'diff'
v2.0.3
```

File: .circleci/config.yml [hunk 1]
```diff
@@ -135,6 +135,7 @@ jobs:
               LAST_COMMIT=`curl -s ${GITHUB_API}repos/${REPO_NAME}/${IMAGE_NAME}/pulls/${PR_NUM} | jq .title`
             fi
             touch ./.tag
+            # Diff to prevent pipeline re-runs.
             if VERSION=`git describe --abbrev=0 --tags` && [ ! -z "`git diff $VERSION`" -o -z "$VERSION" ]; then
```
[the remark this answers is not in the document]
Comment (kyfujisa) [reply .circleci/config.yml:139]: Ugh, it's true, this is weird.
I' ll implement the moves in the following table.


| `git diff $VERSION` | $VERSION | action                                                                            |
| ------------------- | -------- | --------------------------------------------------------------------------------- |
| not null            | not null | release                                                                           |
| null                | not null | don't release (already released)                                                  |
| not null            | null     | first release (treat $VERSION as v0.0.0) (I think it's a non occurring condition) |
| null                | null     | first release (treat $VERSION as v0.0.0)                                          |

[the remark this answers is not in the document]
Comment (WindzCUHK) [reply .circleci/config.yml:139]: Agree. Below script has the similar logic.

**Comment by WindzCUHK** [comment]
`jq` => `jq -r`
fixed: https://github.com/yahoojapan/authorization-proxy/pull/61/files/23a3dbbd5a85ea33dc638f2f5e0ab51ca3fee278..ef95e570552a2b5c2296ae58ca9ace9dd7f8e037

**Pull request merged**